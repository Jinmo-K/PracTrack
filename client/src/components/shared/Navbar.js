import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
// Components
import Collapse from '@material-ui/core/Collapse';
import Fade from '@material-ui/core/Fade';
// Functions
import { logoutUser } from '../../actions/authActions';

const Navbar = ({ auth, logoutUser, history }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [path, setPath] = useState(history.location.pathname);
  const toggleOpen = () => setIsOpen(!isOpen);

  /**
   * Handles logout button click by calling logoutUser function
   * @param {React.MouseEvent} e  The mouse-click event 
   */
  const onLogoutClick = (e) => {
    e.preventDefault();
    logoutUser();
  };

  /**
   * Closes the (mobile) navbar if it is open and user navigates to a different page
   */
  useEffect(() => {
    if (history.location.pathname !== path) {
      setPath(history.location.pathname);
      setIsOpen(false);
    }
  }, [history.location.pathname, path]);

  return (
    <div className='navbar navbar-light navbar-expand-sm sticky-top p-0'>
      <div className='container'>
        <Link to='/' className='navbar-brand d-flex align-items-center p-0 nav-btn'>
          <img src='../../../images/logo-sm.png' alt='logo'></img>
          <strong className='ml-1 mt-1'>PracTrack</strong>
        </Link>
        <button className='navbar-toggler' type='button' onClick={toggleOpen}>
          <span className='navbar-toggler-icon'></span>
        </button>
        <div className='collapse navbar-collapse'>
          {(auth.isAuthenticated)
            ? <ul className='navbar-nav ml-auto'>
                <li className='nav-item'>
                  <Link to='/settings' className='nav-link nav-btn pb-0'>
                    <i className='material-icons'>account_circle</i>
                  </Link>
                </li>
                <li className='nav-item'>
                  <Link to='/' 
                        className='nav-link nav-btn pb-0 logout-btn' 
                        onClick={onLogoutClick}
                  >
                    Logout
                  </Link>
                </li>
              </ul>
            : <ul className='navbar-nav ml-auto'>
                <li className='nav-item'>
                  <Link to='/' className='nav-link nav-btn pb-0'>
                    Login
                  </Link>
                </li>
                <li className='nav-item'>
                  <Link to='/register' className='nav-link nav-btn pb-0'>
                    Signup
                  </Link>
                </li>
              </ul>
          }
        </div>
      </div>
      {/* Dropdown menu for mobile */}
      <div className='d-block d-sm-none w-100' style={{fontWeight: 'bold'}}> 
        <Collapse in={isOpen} collapsedHeight={0} timeout={200}>
          <Fade in={isOpen} timeout={400}>
            <div className='pl-4 pr-2 mb-2'>
              {(auth.isAuthenticated)
              ? <ul className='navbar-nav'>
                  <li className='nav-item align-items-center'>
                    <Link to='/settings' className='nav-link nav-btn pb-0'>
                      Settings
                    </Link>
                  </li>
                  <span className="border" style={{maxWidth: '10ch'}}></span>
                  <li className='nav-item'>
                    <Link to='/' 
                          className='nav-link nav-btn pb-0 logout-btn' 
                          onClick={onLogoutClick}
                    >
                      Logout
                    </Link>
                  </li>
                </ul>
              :
                <ul className='navbar-nav'>
                  <li className='nav-item align-items-center'>
                    <Link to='/' className='nav-link nav-btn pb-0'>
                      Login
                    </Link>
                  </li>
                  <span className="border" style={{maxWidth: '10ch'}}></span>
                  <li className='nav-item'>
                    <Link to='/register' className='nav-link nav-btn pb-0'>
                      Signup
                    </Link>
                  </li>
                </ul>
              }
            </div> 
          </Fade>
        </Collapse>
      </div>
    </div>
  );
}


Navbar.propTypes = {
  /** Function that dispatches logout action */
  logoutUser: PropTypes.func.isRequired,
  /** Auth store state */
  auth: PropTypes.object.isRequired,
  /** Router history */
  history: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({
  auth: state.auth,
});

export default withRouter(connect(
  mapStateToProps,
  { logoutUser }
)(Navbar));

export { Navbar };
