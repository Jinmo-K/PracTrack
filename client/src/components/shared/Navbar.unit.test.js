import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router} from 'react-router-dom';
import { createMemoryHistory } from 'history'
import { Navbar } from './Navbar';

describe('<Navbar />', () => {
  let wrapper;
  // Props
  const logoutUser = jest.fn();
  let history = createMemoryHistory('/');
  const auth_user = {
    isAuthenticated: true
  };
  const auth_visitor = {
    isAuthenticated: false
  };

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = null;
  });

  it('Should render initial layout correctly for users that are not logged in', () => {
    const tree = renderer
      .create(
        <Router>
          <Navbar history={history} auth={auth_visitor} logoutUser={logoutUser}/>
        </Router>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render initial layout correctly for logged in users', () => {
    const tree = renderer
      .create(
        <Router>
          <Navbar history={history} auth={auth_user} logoutUser={logoutUser} />
        </Router>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should dispatch logout action when logout button is clicked', () => {
    wrapper = shallow(<Navbar history={history} auth={auth_user} logoutUser={logoutUser} />);
    wrapper.find('.logout-btn').first().simulate('click', { preventDefault: jest.fn() });
    expect(logoutUser).toBeCalledTimes(1);
  });
});
