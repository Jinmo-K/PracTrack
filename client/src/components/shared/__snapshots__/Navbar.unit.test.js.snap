// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> Should render initial layout correctly for logged in users 1`] = `
<div
  className="navbar navbar-light navbar-expand-sm sticky-top p-0"
>
  <div
    className="container"
  >
    <a
      className="navbar-brand d-flex align-items-center p-0 nav-btn"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="logo"
        src="../../../images/logo-sm.png"
      />
      <strong
        className="ml-1 mt-1"
      >
        PracTrack
      </strong>
    </a>
    <button
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
    >
      <ul
        className="navbar-nav ml-auto"
      >
        <li
          className="nav-item"
        >
          <a
            className="nav-link nav-btn pb-0"
            href="/settings"
            onClick={[Function]}
          >
            <i
              className="material-icons"
            >
              account_circle
            </i>
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            className="nav-link nav-btn pb-0 logout-btn"
            href="/"
            onClick={[Function]}
          >
            Logout
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="d-block d-sm-none w-100"
    style={
      Object {
        "fontWeight": "bold",
      }
    }
  >
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            className="pl-4 pr-2 mb-2"
            style={
              Object {
                "opacity": 0,
                "visibility": "hidden",
              }
            }
          >
            <ul
              className="navbar-nav"
            >
              <li
                className="nav-item align-items-center"
              >
                <a
                  className="nav-link nav-btn pb-0"
                  href="/settings"
                  onClick={[Function]}
                >
                  Settings
                </a>
              </li>
              <span
                className="border"
                style={
                  Object {
                    "maxWidth": "10ch",
                  }
                }
              />
              <li
                className="nav-item"
              >
                <a
                  className="nav-link nav-btn pb-0 logout-btn"
                  href="/"
                  onClick={[Function]}
                >
                  Logout
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Navbar /> Should render initial layout correctly for users that are not logged in 1`] = `
<div
  className="navbar navbar-light navbar-expand-sm sticky-top p-0"
>
  <div
    className="container"
  >
    <a
      className="navbar-brand d-flex align-items-center p-0 nav-btn"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="logo"
        src="../../../images/logo-sm.png"
      />
      <strong
        className="ml-1 mt-1"
      >
        PracTrack
      </strong>
    </a>
    <button
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
    >
      <ul
        className="navbar-nav ml-auto"
      >
        <li
          className="nav-item"
        >
          <a
            className="nav-link nav-btn pb-0"
            href="/"
            onClick={[Function]}
          >
            Login
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            className="nav-link nav-btn pb-0"
            href="/register"
            onClick={[Function]}
          >
            Signup
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="d-block d-sm-none w-100"
    style={
      Object {
        "fontWeight": "bold",
      }
    }
  >
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            className="pl-4 pr-2 mb-2"
            style={
              Object {
                "opacity": 0,
                "visibility": "hidden",
              }
            }
          >
            <ul
              className="navbar-nav"
            >
              <li
                className="nav-item align-items-center"
              >
                <a
                  className="nav-link nav-btn pb-0"
                  href="/"
                  onClick={[Function]}
                >
                  Login
                </a>
              </li>
              <span
                className="border"
                style={
                  Object {
                    "maxWidth": "10ch",
                  }
                }
              />
              <li
                className="nav-item"
              >
                <a
                  className="nav-link nav-btn pb-0"
                  href="/register"
                  onClick={[Function]}
                >
                  Signup
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
